text/3348-c-str-literal.md

```diff
@@ -0,0 +1,127 @@
+  making `cstr!("..")` syntax quite cumbersome.
+
+- No `c""` literal, but make it possible for `""` to implicitly become a `&CStr` through magic.
+
+  We already allow integer literals (e.g. `123`) to become one of many types, so perhaps we could do the same to string literals.
+
+  (It could be a built-in fixed set of types (e.g. just `str`, `[u8]`, and `CStr`),
+  or it could be something extensible through something like a `const trait FromStringLiteral`.
+  Not sure how that would exactly work, but it sounds cool.)
+
+* Allowing only valid UTF-8 and unicode-oriented escape codes (like in `"…"`, e.g. `螃蟹` or `\u{1F980}` but not `\xff`).
+
+  For regular string literals, we have this restriction because `&str` is required to be valid UTF-8.
+  However, C literals (and objects of our `&CStr` type) aren't necessarily valid UTF-8.
+
+* Allowing only ASCII characters rand byte-oriented escape codes (like in `b"…"`, e.g. `\xff` but not `螃蟹` or `\u{1F980}`).
+
+  While C literals (and  `&CStr`) aren't necessarily valid UTF-8, they often do contain UTF-8 data.
+  Refusing to put UTF-8 in it would make the feature less useful and would unnecessarily make it harder to use unicode in programs that mainly use C strings.
+
+* Having separate `c"…"` and `bc"…"` string literal prefixes for UTF-8 and non-UTF8.
+
+  Both of those would be the same type (`&CStr`). Unless we add a special "always valid UTF-8 C string" type, there's not much use in separating them.
+
+* Use `z` instead of `c` (`z"…"`), for "zero terminated" instead of "C string".
+
+  We already have a type called `CStr` for this, so `c` seems consistent.
+
+- Also add `c'…'` as [`c_char`](https://doc.rust-lang.org/stable/core/ffi/type.c_char.html) literal.
+
+  It'd be identical to `b'…'`, except it'd be a `c_char` instead of `u8`.
+
+  This would easily lead to unportable code, since `c_char` is `i8` or `u8` depending on the platform. (Not a wrapper type, but a direct type alias.)
+  E.g. `fn f(_: i8) {} f(c'a');` would compile only on some platforms.
+
+  An alternative is to allow `c'…'` to implicitly be either a `u8` or `i8`. (Just like integer literals can implicitly become one of many types.)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- The `CStr` type needs some work. `&CStr` is currently a wide pointer, but it's supposed to be a thin pointer. See https://doc.rust-lang.org/1.65.0/src/core/ffi/c_str.rs.html#87
+
+  It's not a blocker, but we might want to try to fix that before stabilizing `c"…"`.
+
+# Prior art
+[prior-art]: #prior-art
+
+- C as C string literals (`"…"`). :)
+- Nim has `cstring"…"`.
+- COBOL has `Z"…"`.
+- Probably a lot more languages, but it's hard to search for. :)
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Also add `c'…'` C character literals? (`u8`, `i8`, `c_char`, or something more flexible?)
+
+- Should we make `&CStr` a thin pointer before stabilizing this? (If so, how?)
```

Comment:
I think this should be a blocker on stabilization, yeah.

Reply:
I don't see how this feature is blocked by that at all really, it produces an `&'static CStr` regardless of what `&CStr` itself is made of.

Reply:
@Kixiron To be clear, I think *considering* that question should be a blocker for stabilization.

Given that a major use case of this will be FFI, it seems important that we have a *simple*, not-error-prone way of passing a C string to C functions. If we decide that `&CStr` wasn't that mechanism, then we should decide what that mechanism *should* be, and make sure `c"..."` works well with that.